src/Client/Updater.php

```diff
@@ -644,6 +640,70 @@ public function download(string $target, ...$extra): PromiseInterface
             });
     }
 
+    /**
+     * Gets a target metadata object that contains the specified target.
+     *
+     * @param string $target
+     *   The path of the target file. Needs to be known to the most recent
+     *   targets metadata downloaded in ::refresh().
+     * @param \Tuf\Metadata\TargetsMetadata|null $targetsMetadata
+     *   The targets metadata to search or null. If null then the search will
+     *   start at the top level 'targets.json' file.
+     * @param string[] $searchedRoles
+     *   The roles that have already been searched.
+     *
+     * @return \Tuf\Metadata\TargetsMetadata|null
+     *   The target metadata with a match for the target, or null no match is
+     *   found.
+     */
+    protected function getMetadataForTarget(string $target, ?TargetsMetadata $targetsMetadata = null, array $searchedRoles = []): ?TargetsMetadata
+    {
+        if ($targetsMetadata === null) {
+            // If no target metadata is provided then start searching with the top level targets.json file.
+            /** @var \Tuf\Metadata\TargetsMetadata $targetsMetadata */
+            $targetsMetadata = TargetsMetadata::createFromJson($this->durableStorage['targets.json']);
+            if ($targetsMetadata->hasTarget($target)) {
+                return $targetsMetadata;
+            }
+        }
+
+        $delegatedKeys = $targetsMetadata->getDelegatedKeys();
+        foreach ($delegatedKeys as $keyId => $delegatedKey) {
+            $this->keyDB->addKey($keyId, $delegatedKey);
+        }
+        foreach ($targetsMetadata->getDelegatedRoles() as $delegatedRole) {
+            $delegatedRoleName = $delegatedRole->getName();
+            if (in_array($delegatedRoleName, $searchedRoles, true)) {
+                // TUF-SPEC-v1.0.9 Section 5.4.5.1
+                // If this role has been visited before, then skip this role (so that cycles in the delegation graph are avoided).
+                continue;
+            }
+            if (!$this->roleDB->roleExists($delegatedRoleName)) {
+                $this->roleDB->addRole($delegatedRole);
+            }
+            if (!$delegatedRole->matchesRolePath($target)) {
+                continue;
+            }
+
+            $this->fetchAndVerifyTargetsMetadata($delegatedRoleName);
```

Comment:
Looking at these two lines, it feels like `fetchAndVerifyTargetsMetadata` should return an instance of `TargetsMetadata`.